### PR TITLE
Refactor `orchest-controller` to use State-machine Pattern and Add tests

### PR DESCRIPTION
**Pull request by nhaghighat**
## Description

A video is posted on notion, which explains the new architecture, will also add a README

Fixes: #issue

## Checklist

- [ ] I have manually tested my changes and I am happy with the result.
- [ ] The documentation reflects the changes.
- [ ] Add detailed README about the controller architecture and a state diagram
- [ ] Add webhooks
- [ ] The PR branch is set up to merge into `dev` instead of `master`.

### CODE REVIEW

File: services/orchest-controller/pkg/addons/components.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+)
+
+func RegisterAddon(name string, addon Addon) {
+	if _, ok := Registry.addons[name]; !ok {
+		Registry.addons[name] = addon
+		klog.Infof("Addon %s is registered with the registry", name)
+		return
+	}
+
+	klog.Infof("An Addon with the same name is already registered with the registry, name=%s", name)
```
Comment (fruttasecca) [inline services/orchest-controller/pkg/addons/components.go]: Should this be an error?

File: services/orchest-controller/pkg/controller/orchestcluster/cluster_utils.go [hunk 2]
```diff
@@ -24,15 +23,12 @@ import (
...
-		controller.AuthServer,
-		controller.OrchestWebserver,
-		controller.BuildKitDaemon,
-		controller.NodeAgent,
+	deploymentStages = [][]string{
+		{controller.OrchestDatabase, controller.Rabbitmq},
+		{controller.OrchestApi},
+		{controller.CeleryWorker},
+		{controller.AuthServer},
+		{controller.OrchestWebserver, controller.BuildKitDaemon, controller.NodeAgent},
```
Comment (fruttasecca) [inline services/orchest-controller/pkg/controller/orchestcluster/cluster_utils.go:29]: Not sure if matters that much but note that if the webserver becomes available before the buildkit daemon is up (or, at a higher abstraction level, before everything is really ready) users could run into funny issues.
Moreover, I think that things can be made a bit more parallel:
- db, rabbitmq, buildkit daemon
- orchest-api, auth-server
- celery-worker, node-agent
- webserver

File: services/orchest-controller/pkg/controller/orchestcluster/state_machine.go [hunk 3]
```diff
@@ -15,6 +15,11 @@ import (
 	"k8s.io/klog/v2"
 )
 
+var (
+	deployTimeOut = time.Minute * 5
```
Comment (fruttasecca) [inline services/orchest-controller/pkg/controller/orchestcluster/state_machine.go:19]: Is this actually in use? I'm not sure I'd like a slow image pull to lead to errors

File: services/orchest-controller/pkg/controller/orchestcluster/running_state.go [hunk 4]
```diff
@@ -17,5 +17,9 @@ func (state *RunningState) To(ctx context.Context, stateMachine *OrchestStateMac
 }
 
 func (state *RunningState) Do(ctx context.Context, stateMachine *OrchestStateMachine, orchest *orchestv1alpha1.OrchestCluster) error {
+	if orchest.Status.ObservedGeneration != orchest.Generation {
+		// If the ObservedGeneration is different that the actual Generation we go to stopping state.
+		return stateMachine.toState(ctx, orchestv1alpha1.Stopping)
+	}
```
Comment (fruttasecca) [inline services/orchest-controller/pkg/controller/orchestcluster/running_state.go:29]: I'm a bit confused by this, could you clarify?

**Comment by fruttasecca** [comment]
I really like that we are moving to components and a state machine, and the video was really useful. Aside from that, my suggestion for the next time would be to split the PR into stacked PRs to make it more reviewable. I'm also really looking forward to the README and/or a diagram of the state machine.

With this said, is the PR to be considered complete? I.e. should things work? I've tried an installation a bunch of times and it seems to always fail when deploying the registry
```
│ I1027 10:34:44.330883       1 state_machine.go:189] Creating docker-registry, namespace=orchest, name=cluster-1 retry=5              │
│ I1027 10:34:44.439002       1 state_machine.go:189] Creating argo-workflow, namespace=orchest, name=cluster-1 retry=5                │
│ I1027 10:34:46.426914       1 state_machine.go:189] Creating docker-registry, namespace=orchest, name=cluster-1 retry=4              │
│ I1027 10:34:48.647449       1 state_machine.go:189] Creating docker-registry, namespace=orchest, name=cluster-1 retry=3              │
│ I1027 10:34:51.043650       1 state_machine.go:189] Creating docker-registry, namespace=orchest, name=cluster-1 retry=2              │
│ I1027 10:34:53.240732       1 state_machine.go:189] Creating docker-registry, namespace=orchest, name=cluster-1 retry=1              │
│ I1027 10:34:56.240936       1 state_machine.go:322] Transition from Deploying Third Parties to Error state, name=cluster-1, namespac │
│ e=orchest                                                                                                                            │
│ I1027 10:34:57.043362       1 state_machine.go:322] Transition from Error to Stopping state, name=cluster-1, namespace=orchest       │
│ I1027 10:34:57.043467       1 state_machine.go:252] Deleting orchest-webserver, namespace=orchest, name=cluster-1                    │
│ I1027 10:34:57.640685       1 state_machine.go:252] Deleting node-agent, namespace=orchest, name=cluster-1                           │
│ I1027 10:35:00.645879       1 state_machine.go:252] Deleting auth-server, namespace=orchest, name=cluster-1                          │
│ I1027 10:35:02.231033       1 state_machine.go:252] Deleting celery-worker, namespace=orchest, name=cluster-1                        │
│ I1027 10:35:03.643590       1 state_machine.go:252] Deleting orchest-api, namespace=orchest, name=cluster-1                          │
│                                                                                                                                      │

```

**Comment by fruttasecca** [comment]
Given that we now have controller tests we should probably add those to the `scripts/run_tests.sh` and re-enable the GH workflow running that (`.github/workflows/python-tests.yml`) , but only having the controller tests as part of the arguments matrix. I guess a rename of the workflow is also in order.

**Comment by fruttasecca** [comment]
The `services/orchest-controller/pkg/components/helm_componenet.go` name needs fixing

**Comment by fruttasecca** [comment]
There is a mismatch between the orchest phases in `services/orchest-controller/pkg/apis/orchest/v1alpha1/types.go` and the states of the state machine, i.e. the state machine has less states. Is it because some phases are treated by the same state? If so, might be worth to have a diagram mapping phases to states.

When it comes to the state machine state and the orchest cluster phase/CRD, what was the reasoning behind not having the state machine state be solely defined by the CRD? What I'm afraid of is an inconsistency between the two states, but I might be missing the full picture

**Comment by fruttasecca** [comment]
Speaking of a README and or states diagram, I think the `services/orchest-controller/pkg/controller/orchestcluster/state_machine.go` file could also be a good place for that information

File: services/orchest-controller/pkg/controller/orchestcluster/running_state.go (hunk 4 above)
Comment (nhaghighat) [reply services/orchest-controller/pkg/controller/orchestcluster/running_state.go:29]: Once we go to the running state, the `ObservedGeneration` get updated, once a user changes the CR, the `Generation` of the CR will change, and if these numbers are not the same, we realize we should update. and update is nothing more than stop, then start with new configs.

**Comment by nhaghighat** [comment]
> I really like that we are moving to components and a state machine, and the video was really useful. Aside from that, my suggestion for the next time would be to split the PR into stacked PRs to make it more reviewable. I'm also really looking forward to the README and/or a diagram of the state machine.

Fixed this issue, and will add a readme.

**Comment by nhaghighat** [comment]
> There is a mismatch between the orchest phases in services/orchest-controller/pkg/apis/orchest/v1alpha1/types.go and the states of the state machine, i.e. the state machine has fewer states. Is it because some phases are dealt with by the same state? If so, might be worth having a diagram mapping phases to states.

Will remove the extra, unused states, for example, Staring state and Deploying State, are practically the same, so Starting is removed.

**Comment by fruttasecca** [comment]
Just realized this PR also introduces the nvidia device plugin, wasn't that part of another PR? :thinking: Can't find that PR, was it decided to merge those changes regardless?

**Comment by ricklamers** [comment]
We’re postponing this refactor for the time being 🙌🏻

**Pull request closed without merging**